openpype/hosts/houdini/api/plugin.py

```diff
@@ -84,3 +90,116 @@ def process(self):
                 OpenPypeCreatorError,
                 OpenPypeCreatorError("Creator error: {}".format(er)),
                 sys.exc_info()[2])
+
+
+@six.add_metaclass(ABCMeta)
+class HoudiniCreator(NewCreator):
+    selected_nodes = []
+
+    @staticmethod
+    def _create_instance_node(
+            node_name, parent,
+            node_type="geometry"):
+        # type: (str, str, str) -> hou.Node
+        """Create node representing instance.
+
+        Arguments:
+            node_name (str): Name of the new node.
+            parent (str): Name of the parent node.
+            node_type (str, optional): Type of the node.
+
+        Returns:
+            hou.Node: Newly created instance node.
+
+        """
+        parent_node = hou.node(parent)
+        instance_node = parent_node.createNode(
+            node_type, node_name=node_name)
+        instance_node.moveToGoodPosition()
+        return instance_node
+
+    def create(self, subset_name, instance_data, pre_create_data):
+        try:
+            if pre_create_data.get("use_selection"):
+                self.selected_nodes = hou.selectedNodes()
+
+            # Get the node type and remove it from the data, not needed
+            node_type = instance_data.pop("node_type", None)
+            if node_type is None:
+                node_type = "geometry"
+
+            instance_node = self._create_instance_node(
+                subset_name, "/out", node_type)
+
+            # wondering if we'll ever need more than one member here
+            # in Houdini
+            # instance_data["members"] = [instance_node.path()]
+            instance_data["instance_node"] = instance_node.path()
+
+            instance = CreatedInstance(
+                self.family,
+                subset_name,
+                instance_data,
+                self)
+            self._add_instance_to_context(instance)
+            imprint(instance_node, instance.data_to_store())
+            return instance
+
+        except hou.Error as er:
+            six.reraise(
+                OpenPypeCreatorError,
+                OpenPypeCreatorError("Creator error: {}".format(er)),
+                sys.exc_info()[2])
+
+    def lock_parameters(self, node, parameters):
+        """Lock list of specified parameters on the node.
+
+        Args:
+            node (hou.Node): Houdini node to lock parameters on.
+            parameters (list of str): List of parameter names.
+
+        """
+        for name in parameters:
+            try:
+                parm = node.parm(name)
+                parm.lock(True)
+            except AttributeError:
+                self.log.debug("missing lock pattern {}".format(name))
+
+    def collect_instances(self):
+        cached_instances = self.collection_shared_data.get(
```

Comment:
It was meant more like this:
```
if "houdini_cached_instances" not in self.collection_shared_data:
    instances_by_identifier = collections.defaultdict(list)
    for instance in all_instances:
        identifier = instance.get("creator_identifier")
        if identifier:
            instances_by_identifier[identifier].append(instance)

    self.collection_shared_data["houdini_cached_instances"] = (
        instances_by_identifier
    )

instances = (
    self.collection_shared_data
    ["houdini_cached_instances"]
    .get(self.identifier)
) or []
for instance in instances:
    created_instance = CreatedInstance.from_existing(
        read(instance), self
    )
    self._add_instance_to_context(created_instance)
```
So the `list_instances` is called only once.

Reply:
isn't it doing exactly the same? In pseudo code:

```
get cached instances for given creator
if nothing instances:
    query the scene
     fill the cache
```

Reply:
With current implementation you can skip the caching completelly. Each creator will call `list_instances` and store only instances for him -> that does not need caching at all.

Reply:
now I am little lost. So how do I get the list of instances for `collect_instances()`? Is it done automatically now?

Reply:
The shared data are meant as "Shared data" across multiple creators for `collect_instances`. If each creator will cache data only for the creator it is pointless.

The goal is for example to collect all nodes in scene with `id` which is `pyblish.avalon.instance` -> store these to shared data so other creators don't have to collect the nodes again (instead of going through 1.000.000 nodes every time go through them only once and store relevant nodes for other creators).